### PR TITLE
netreg: add iterator

### PR DESCRIPTION
**Pull request by authmillenon**
A minor readability enhancement, inspired by `utlist`

### CODE REVIEW

**Comment by haukepetersen** [comment]
NACK for two reasons: 
  1. I think the use of macros in API code should be kept to the absolute minimum!
  2. With this you expose internals of the netreg implementation to the API, what happens if you decide to implement netreg internally using something else the `utlist`?

**Comment by authmillenon** [comment]
This does neither utilize `utlist` (I just got inspired by it's `LL_FOREACH`) nor does it expose the API... I use publicly defined functions for that (have a look at the code please). On function-like macros: technically this is no function-like macro but an iterator macro and I find them extremely useful, because of a 3-line for-loop head I have a very short loop head.

File: sys/net/link_layer/ng_nomac/ng_nomac.c [hunk 1]
```diff
@@ -40,22 +40,20 @@ static void _event_cb(ng_netdev_event_t event, void *data)
     if (event == NETDEV_EVENT_RX_COMPLETE) {
         ng_pktsnip_t *pkt;
         ng_netreg_entry_t *sendto;
+        int sendto_num = ng_netreg_num(pkt->type, NG_NETREG_DEMUX_CTX_ALL);
```
Comment (haukepetersen) [inline sys/net/link_layer/ng_nomac/ng_nomac.c]: does not work: pkt is undefined here...

Comment (authmillenon) [inline sys/net/link_layer/ng_nomac/ng_nomac.c]: Fixed

**Comment by haukepetersen** [comment]
so we are talking
```c
int sendto_num;

/* get pointer to the received packet */
pkt = (ng_pktsnip_t *)data;
/* find out, who to send the packet to */
sendto_num = ng_netreg_num(pkt->type, NG_NETREG_DEMUX_CTX_ALL);
/* throw away packet if no one is interested */
if (sendto_num == 0) {
    DEBUG("nomac: unable to forward packet of type %i\n", pkt->type);
    ng_pktbuf_release(pkt);
}
/* send the packet to everyone interested in it's type */
ng_pktbuf_hold(pkt, sendto_num - 1);
NG_NETREG_FOREACH(sendto, pkt->type, NG_NETREG_DEMUX_CTX_ALL) {
    DEBUG("nomac: sending pkt %p to PID %u\n", pkt, sendto->pid);
    ng_netapi_receive(sendto->pid, pkt);
}
```
vs.
```c
int sendto_num;
ng_netreg_entry_t *sendto;
 
/* get pointer to the received packet */
pkt = (ng_pktsnip_t *)data;
/* find out, who to send the packet to */
sendto_num = ng_netreg_num(pkt->type, NG_NETREG_DEMUX_CTX_ALL);
/* throw away packet if no one is interested */
if (sendto_num == 0) {
    DEBUG("nomac: unable to forward packet of type %i\n", pkt->type);
    ng_pktbuf_release(pkt);
}
/* send the packet to everyone interested in it's type */
sendto = ng_netreg_lookup(pkt->type, NG_NETREG_DEMUX_CTX_ALL);
ng_pktbuf_hold(pkt, sendto_num - 1);
while (sendto != NULL) {
    DEBUG("nomac: sending pkt %p to PID %u\n", pkt, sendto->pid);
    ng_netapi_receive(sendto->pid, pkt);
    sendto = ng_netreg_getnext(sendto);
}
```
By identical code size. I am clearly in favor of solution 2! We are programming C afer all, and I am no fan what so ever in trying to simulate other high level languages as Java or C# by introducing strange looking shortcut macros!

**Comment by authmillenon** [comment]
Okay... But one of the core features of C and C++ was always people misusing their freedom ;P

**Pull request closed without merging**